src/test/java/com/puppycrawl/tools/checkstyle/internal/ImmutabilityTest.java

```diff
@@ -0,0 +1,469 @@
+            + ".IGNORED_TYPES",
+        "com.puppycrawl.tools.checkstyle.checks.coding.InnerAssignmentCheck"
+            + ".ALLOWED_ASSIGNMENT_CONTEXT",
+        "com.puppycrawl.tools.checkstyle.checks.coding.InnerAssignmentCheck"
+            + ".ALLOWED_ASSIGNMENT_IN_COMPARISON_CONTEXT",
+        "com.puppycrawl.tools.checkstyle.checks.coding.InnerAssignmentCheck.COMPARISON_TYPES",
+        "com.puppycrawl.tools.checkstyle.checks.coding.InnerAssignmentCheck.CONTROL_CONTEXT",
+        "com.puppycrawl.tools.checkstyle.checks.coding.InnerAssignmentCheck"
+            + ".LOOP_IDIOM_IGNORED_PARENTS",
+        "com.puppycrawl.tools.checkstyle.checks.coding.MatchXpathCheck.xpathExpression",
+        "com.puppycrawl.tools.checkstyle.checks.javadoc.AtclauseOrderCheck.DEFAULT_ORDER",
+        "com.puppycrawl.tools.checkstyle.checks.javadoc.JavadocBlockTagLocationCheck.DEFAULT_TAGS",
+        "com.puppycrawl.tools.checkstyle.checks.javadoc.SummaryJavadocCheck.ALLOWED_TYPES",
+        "com.puppycrawl.tools.checkstyle.checks.modifier.ModifierOrderCheck.JLS_ORDER",
+        "com.puppycrawl.tools.checkstyle.checks.modifier.RedundantModifierCheck"
+            + ".TOKENS_FOR_INTERFACE_MODIFIERS",
+        "com.puppycrawl.tools.checkstyle.checks.regexp.RegexpMultilineCheck.detector",
+        "com.puppycrawl.tools.checkstyle.checks.regexp.RegexpSinglelineCheck.detector",
+        "com.puppycrawl.tools.checkstyle.checks.coding.IllegalTokenTextCheck.formatString",
+        "com.puppycrawl.tools.checkstyle.checks.javadoc.WriteTagCheck.tagRegExp",
+        "com.puppycrawl.tools.checkstyle.checks.naming.AbstractNameCheck.format",
+        "com.puppycrawl.tools.checkstyle.checks.whitespace.AbstractParenPadCheck.option"
+    );
+
+    /**
+     * List of classes not following
+     * {@link #testClassesWithImmutableFieldsShouldBeStateless()} rule.
+     */
+    private static final Set<String> SUPPRESSED_CLASSES_FOR_STATELESS_CHECK_RULE = Set.of(
+        "com.puppycrawl.tools.checkstyle.checks.javadoc.JavadocPackageCheck",
+        "com.puppycrawl.tools.checkstyle.checks.javadoc.MissingJavadocMethodCheck",
+        "com.puppycrawl.tools.checkstyle.checks.metrics.ClassDataAbstractionCouplingCheck",
+        "com.puppycrawl.tools.checkstyle.checks.metrics.ClassFanOutComplexityCheck",
+        "com.puppycrawl.tools.checkstyle.checks.naming.CatchParameterNameCheck",
+        "com.puppycrawl.tools.checkstyle.checks.naming.ClassTypeParameterNameCheck",
+        "com.puppycrawl.tools.checkstyle.checks.naming.ConstantNameCheck",
+        "com.puppycrawl.tools.checkstyle.checks.naming.InterfaceTypeParameterNameCheck",
+        "com.puppycrawl.tools.checkstyle.checks.naming.LambdaParameterNameCheck",
+        "com.puppycrawl.tools.checkstyle.checks.naming.LocalFinalVariableNameCheck",
+        "com.puppycrawl.tools.checkstyle.checks.naming.LocalVariableNameCheck",
+        "com.puppycrawl.tools.checkstyle.checks.naming.MemberNameCheck",
```

Comment:
Just a conversation while waiting for the last item to be done ( https://github.com/checkstyle/checkstyle/pull/12135#discussion_r968698797 ),

> MemberNameCheck

I took a look at this, and this class has no annotation on it. It extends `AbstractAccessControlNameCheck` which also has no annotation on it. It also extends `AbstractNameCheck` which has `@StatelessCheck` on it.

Is this violation solely because the annotation is on parent class, which is 2 level deep?

Reply:
Violation is because the class itself has immutable fields (no fields in this case, not the inherited fields) and has no annotation.

Reply:
It has an annotation but not the main class itself. I assume that via reflection the annotation will show up, but I was curious if ArchUnit was not seeing the annotation?

Reply:
Yes, ArchUnit is not seeing it, it also has `@Inherited`, I will report this to ArchUnit team.